### PR TITLE
plat/kvm: Remove intermediate relocatable object step

### PR DESCRIPTION
**Pull request by jschlumpp**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): `kvm`
 - Application(s): N/A

### Description of changes

This object was previously used to hide all symbols except the entry point. But this step was removed making the relocatable object creation unnecessary.

### CODE REVIEW

**Review by mogasergiu — COMMENTED** [review]
Would it be possible to do the same for Xen? Something like:
```
--- a/plat/xen/Linker.uk
+++ b/plat/xen/Linker.uk
@@ -20,8 +20,8 @@ XEN_LD_SCRIPT_FLAGS += $(addprefix -Wl$(comma)-T$(comma),\
 
 $(XEN_DEBUG_IMAGE): $(XEN_ALIBS) $(XEN_ALIBS-y) $(XEN_OLIBS) $(XEN_OLIBS-y) \
 		    $(UK_ALIBS) $(UK_ALIBS-y) $(UK_OLIBS) $(UK_OLIBS-y)
-	$(call build_cmd,LD,,$(XEN_IMAGE).ld.o,\
-	       $(LD) -r $(LIBLDFLAGS) $(LIBLDFLAGS-y) \
+	$(call build_cmd,LD,,$(XEN_IMAGE).o,\
+	       $(LD) \
 			$(XEN_LDFLAGS) $(XEN_LDFLAGS-y) \
 			$(XEN_OLIBS) $(XEN_OLIBS-y) \
 			$(UK_OLIBS) $(UK_OLIBS-y) \
@@ -29,14 +29,12 @@ $(XEN_DEBUG_IMAGE): $(XEN_ALIBS) $(XEN_ALIBS-y) $(XEN_OLIBS) $(XEN_OLIBS-y) \
 			$(XEN_ALIBS) $(XEN_ALIBS-y) \
 			$(UK_ALIBS) $(UK_ALIBS-y) \
 			-Wl$(comma)--end-group \
-			-o $(XEN_IMAGE).ld.o)
+                      $(LDFLAGS) $(LDFLAGS-y) \
+		        $(XEN_LD_SCRIPT_FLAGS) \
+		        -o $(XEN_IMAGE).o)
 	$(call build_cmd,OBJCOPY,,$(XEN_IMAGE).o,\
 		$(OBJCOPY) -w -G xenos_* -G _libxenplat_start \
-			$(XEN_IMAGE).ld.o $(XEN_IMAGE).o)
-	$(call build_cmd,LD,,$@,\
-	       $(LD) $(LDFLAGS) $(LDFLAGS-y) $(XEN_LDFLAGS) $(XEN_LDFLAGS-y) \
-		     $(XEN_LD_SCRIPT_FLAGS) \
-		     $(XEN_IMAGE).o -o $@)
+			$(XEN_IMAGE).o $@)
 
 $(XEN_IMAGE): $(XEN_DEBUG_IMAGE)
 	$(call build_cmd,SCSTRIP,,$@,\
```

**Review by mogasergiu — APPROVED** [review]
My, perhaps irrelevant to this KVM PR, Xen related comment aside, this looks good.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

**Comment by jschlumpp** [comment]
There are some warnings related to the ukdebug tracepoints. I have to take another look at this PR soon.

**Comment by mogasergiu** [comment]
> There are some warnings related to the ukdebug tracepoints. I have to take another look at this PR soon.

Hmm, good point, I did not test this with the tracepoint configs set.

File: plat/kvm/Linker.uk [hunk 1]
```diff
@@ -41,12 +41,9 @@ $(KVM_DEBUG_IMAGE): $(KVM_ALIBS) $(KVM_ALIBS-y) $(KVM_OLIBS) $(KVM_OLIBS-y) \
...
-	       $(LD) $(LDFLAGS) $(LDFLAGS-y) \
-		     $(KVM_LDFLAGS) $(KVM_LDFLAGS-y) \
-		     $(KVM_LD_SCRIPT_FLAGS) \
-		     $(KVM_IMAGE).o -o $@)
+			$(LDFLAGS) $(LDFLAGS-y) \
+			$(KVM_LD_SCRIPT_FLAGS) \
+			-o $@)
 	$(call build_bootinfo,$@)
 
 $(KVM_IMAGE): $(KVM_IMAGE).dbg
```
Comment (Krechals) [inline plat/kvm/Linker.uk:51]: We won't need this step anymore:
```
LIBKVMPLAT_CLEAN += $(call build_clean,$(KVM_IMAGE).o)
```

**Review by Krechals — COMMENTED** [review]
I had some KVM builds with `app-helloworld` / `app-httpreply` and there weren't any warnings (not even when tracepoints were enabled). Running also seemed to be fine.

However, I looked a bit into the ELF sections to check if there are any meaningful changes. Based on an `app-httpreply` config, the `.rodata` section seems to have a lower size when removing the linking with the relocatable object.
Before:
```
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .text             PROGBITS         0000000000100000  00100000
       000000000004b77b  0000000000000000  AX       0     0     4096
  [ 2] .rodata           PROGBITS         000000000014c000  0014c000
       0000000000005628  0000000000000000   A       0     0     32
  [ 3] netdev__param_arg PROGBITS         0000000000151628  00151628
```

Now:
```
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .text             PROGBITS         0000000000100000  00100000
       000000000004b77b  0000000000000000  AX       0     0     4096
  [ 2] .rodata           PROGBITS         000000000014c000  0014c000
       000000000000561d  0000000000000000   A       0     0     32
  [ 3] netdev__param_arg PROGBITS         0000000000151620  00151620
       0000000000000048  0000000000000000   A       0     0     8
```

There aren't any symbols from the previous version that aren't included in the new one, so I assume the difference comes only because of symbol alignment.

Still, in this version there are symbols placed in `.rodata` that were previously in `netdev__param_arg`:
- `_uk_thread_inittab_end`
- `uk_fslist_end`
- `uk_fslist_start`

Based on the linker scripts `lib/vfscore/extra.ld` and `lib/uksched/extra.ld`, I would assume these symbols shouldn't be in `.rodata`.

`lib/vfscore/extra.ld`:
```
SECTIONS
{
        .uk_fs_list : {
                PROVIDE(uk_fslist_start = .);
                KEEP (*(.uk_fs_list))
                PROVIDE(uk_fslist_end = .);
        }
}
INSERT AFTER .rodata;
```

**Comment by jschlumpp** [comment]
@Krechals 
> Still, in this version there are symbols placed in `.rodata` that were previously in `netdev__param_arg`:
> 
> * `_uk_thread_inittab_end`
> * `uk_fslist_end`
> * `uk_fslist_start`

I can't reproduce that and these symbols seem to be broken in both cases anyway. Note that `*_start == *_end` would indicate the list is empty and then the linker seems to move the symbols around in strange way. Also the end address is not inclusive.

The warnings I mentioned will be fixed by another PR, which is not published, yet.

**Comment by razvand** [comment]
@skuenzer, please see this PR. Approve it, if all is OK.

**Review by skuenzer — APPROVED** [review]
> @skuenzer, please see this PR. Approve it, if all is OK.

Looks good to me. The mentioned warning does not appear anymore. :+1:

Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request closed without merging**